lib/Driver/Driver.cpp

```diff
@@ -1126,10 +1126,10 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
     // top-level output.
     OI.ShouldGenerateModule = true;
     OI.ShouldTreatModuleAsTopLevelOutput = true;
-  } else if ((OI.DebugInfoKind > IRGenDebugInfoKind::LineTables &&
-              OI.shouldLink()) ||
-             Args.hasArg(options::OPT_emit_objc_header,
-                         options::OPT_emit_objc_header_path)) {
+  } else if (Args.hasArg(options::OPT_emit_objc_header,
+                         options::OPT_emit_objc_header_path) ||
+             (OI.shouldLink() &&
+              TC.getTriple().getObjectFormat() == llvm::Triple::ELF)) {
```

Comment:
> Generating a non-top level module is still required to preserve implicit module wrapping of modules whose objects are ELF.

I'm pretty sure this feature is still only for debugging; it's just how we do that on Linux where dSYMs aren't the usual mechanism. So you should be able to drop this.